### PR TITLE
MINOR: Github build enable Gradle cache and add checkstyle annotations 

### PR DESCRIPTION
**Pull request by mumrah**
Passing --build-cache to gradlew is not enough, we also need to have the build cache enabled in settings.gradle. Once on trunk, this should allow the build-cache to get cached by GitHub. Hopefully this allows actual Gradle build graph, task outputs, and compiled classes to get cached for faster builds.

### CODE REVIEW

**Comment by mumrah** [comment]
This PR verifies two things:

* PRs from committers do not need approval
* The "gh-*" branch prefix is working

**Comment by mumrah** [comment]
In this PR's build I see:
```
Using local directory build cache for the root build (location = /home/runner/.gradle/caches/build-cache-1, removeUnusedEntriesAfter = 7 days).
...
Build cache key for task ':generator:compileJava' is 5a17d58c341b8df5cc4bbe83cc6ee460
```

whereas other PRs, I see:

```
> Task :generator:compileJava
Caching disabled for task ':generator:compileJava' because:
  Build cache is disabled
Task ':generator:compileJava' is not up-to-date because:
  No history is available.
The input changes require a full rebuild for incremental task ':generator:compileJava'.
```

**Comment by gharris1727** [comment]
Is build caching a good idea? Our current jenkins runs `clean`, and I've observed that even that does not properly clean up all state and can lead to unrecoverable builds #16839.

**Comment by mumrah** [comment]
@gharris1727 the Github action runners are VMs that aren't reused, so every build should be starting from a clean slate. If we ever use self-hosted runners, we would need to take care that we clean up the workspace since those can be re-used.

The Gradle cache holds a lot more than just our generated files. If you take a look here https://github.com/apache/kafka/actions/caches, there are 600+ MB of dependencies downloaded on each build that can restored from a cache. I think it this can help our build times.

This whole effort is still experimental, but my aim is to let `trunk` builds populate the cache and PRs restore that cache. If it works as intended, it will allow PRs to have `UP-TO-DATE` and `FROM-CACHE` tasks for things that haven't changed.

File: .github/workflows/pr.yml
Comment (chia7712) [inline .github/workflows/pr.yml]: not sure why we need to set `GITHUB_REPOSITORY` and `GITHUB_BRANCH`

File: settings.gradle
Comment (chia7712) [inline settings.gradle:50]: could you please add comment for this change?

File: .github/workflows/pr.yml
Comment (chia7712) [inline .github/workflows/pr.yml]: Disabling the `spotlessCheck` may misdirect us to merge a PR with imports error. The "import order" check of `spotless` is different from `checkstyle`. `spotless` has stricter check than `checkstyle`

File: .github/workflows/pr.yml
Comment (chia7712) [inline .github/workflows/pr.yml:53]: happy to know the purpose of `--info`

File: .github/scripts/checkstyle.py
Comment (mumrah) [inline .github/scripts/checkstyle.py:64]: This is the GitHub magic ✨. If we print things like "::notice ... ::" to the stdout, it will create annotations.

**Comment by mumrah** [comment]
@chia7712 this PR grew a bit today :) I wanted to try processing the checkstyle report to generate some useful output in the job summary. After a bit of reading, I found it was actually possible to create inline annotations for the checkstyle errors.

Here's an example https://github.com/apache/kafka/pull/16948/files#diff-77dc2adb187fd078084644613cff2b53021c8a5fbcdcfa116515734609d1332a

In case it disappears, here's a screenshot 
image

Pretty neat!

This patch now has the one-liner to enable the Gradle cache as well as this new checkstyle processing. PTAL

File: .github/scripts/checkstyle.py
Comment (chia7712) [inline .github/scripts/checkstyle.py:73]: I do love this script, but have you considered using existent Github action? like https://github.com/jwgmeligmeyling/checkstyle-github-action

**Review by chia7712 — COMMENTED** [review]
@mumrah thanks for this great patch. BTW, could you please update the topic according to changes :)

File: .github/scripts/checkstyle.py
Comment (mumrah) [reply .github/scripts/checkstyle.py:73]: Yes, I did look through the existing marketplace actions before writing this script. We must be cautious when bringing in 3rd party actions (actions from non-verified authors). ASF Infra team has to manually audit each 3rd party action that gets included (similar to the review process for a Jenkins plugin). There was one verified action https://github.com/marketplace/check-run-reporter which seems to cost money. 

Going the custom route is also based on my experience when trying to process JUnit tests with a marketplace action -- it wasn't able to handle the number of tests we have (kept running out of memory).

I figured this was a simple enough task that we could just roll our own script 😄

File: .github/workflows/pr.yml
Comment (mumrah) [reply .github/workflows/pr.yml]: Ah, that's left over from older version. Good catch!

File: .github/workflows/pr.yml
Comment (mumrah) [reply .github/workflows/pr.yml]: Good to know. I've backed out this change already. Do you know if we are planning on keeping both checkstyle and spotless around for a while? Or are we trying to migrated only to spotless?

File: .github/scripts/checkstyle.py
Comment (chia7712) [reply .github/scripts/checkstyle.py:73]: @mumrah thanks for sharing. I guess we need to have custom parser for spotbug report too, right?

File: .github/workflows/pr.yml
Comment (chia7712) [reply .github/workflows/pr.yml]: > Do you know if we are planning on keeping both checkstyle and spotless around for a while? Or are we trying to migrated only to spotless?

Yes, I prefer to leverage spotless as much as possible as it can fix the code automatically (for example, re-format the imports). We will dig in it after dropping JDK8. there is already a related jira https://issues.apache.org/jira/browse/KAFKA-16896

File: .github/workflows/pr.yml
Comment (chia7712) [inline .github/workflows/pr.yml]: Could you please add comments for this magic `1` :smile:

File: .github/scripts/checkstyle.py
Comment (mumrah) [reply .github/scripts/checkstyle.py:73]: Yes, and JUnit eventually. I'll take a look at those next. I suspect it will be quite similar to the checkstyle parser (glob the right path, streaming xml, etc).

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**
